server/routes/basicDetails.ts

```diff
@@ -0,0 +1,411 @@
+      const basicDetailsDateOfLetter: string = req.body.dateOfLetter
+      res.render(`pages/basic-details`, {
+        errorMessages,
+        breachNotice,
+        basicDetails,
+        defaultOffenderAddress,
+        defaultReplyAddress,
+        alternateAddressOptions,
+        replyAddressOptions,
+        basicDetailsDateOfLetter,
+      })
+    } else {
+      // mark that a USER has saved the document at least once
+      breachNotice.basicDetailsSaved = true
+      await breachNoticeApiClient.updateBreachNotice(breachNoticeId, breachNotice)
+      console.log(req.body)
+      console.log(`action is ${req.body.action}`)
+
+      if (req.body.action === 'viewDraft') {
+        try {
+          await showDraftPdf(breachNotice.id, res)
+        } catch (err) {
+          // Render the page with the new error
+          errorMessages.pdfRenderError = {
+            text: 'There was an issue generating the draft report. Please try again or contact support.',
+          }
+
+          const alternateAddressOptions = initiateAlternateAddressSelectItemList(basicDetails, breachNotice)
+          const replyAddressOptions = initiateReplyAddressSelectItemList(basicDetails, breachNotice)
+          const basicDetailsDateOfLetter: string = req.body.dateOfLetter
+          res.render(`pages/basic-details`, {
+            errorMessages,
+            breachNotice,
+            basicDetails,
+            defaultOffenderAddress,
+            defaultReplyAddress,
+            alternateAddressOptions,
+            replyAddressOptions,
+            basicDetailsDateOfLetter,
+          })
+        }
+      } else if (req.body.action === 'saveProgressAndClose') {
+        res.setHeader('Content-Security-Policy', "script-src-elem 'unsafe-inline'")
+        res.send('<script>window.close();</script>')
+      } else {
+        res.redirect(`/warning-type/${req.params.id}`)
+      }
+    }
+  })
+
+  function validateBasicDetails(breachNotice: BreachNotice, userEnteredDateOfLetter: string): ErrorMessages {
+    const errorMessages: ErrorMessages = {}
+    const currentDateAtStartOfTheDay: LocalDateTime = LocalDate.now().atStartOfDay()
+    if (breachNotice.referenceNumber) {
+      if (breachNotice.referenceNumber.trim().length > 30) {
+        errorMessages.officeReferenceNumber = {
+          text: 'The Office Reference entered is greater than the 30 character limit.',
+        }
+      }
+    }
+
+    try {
+      // eslint-disable-next-line no-param-reassign
+      breachNotice.dateOfLetter = fromUserDate(userEnteredDateOfLetter)
+    } catch (error: unknown) {
+      // eslint-disable-next-line no-param-reassign
+      breachNotice.dateOfLetter = userEnteredDateOfLetter
+      errorMessages.dateOfLetter = {
+        text: 'The proposed date for this letter is in an invalid format, please use the correct format e.g 17/5/2024',
+      }
+      // we cant continue with date validation
+      return errorMessages
+    }
+
+    if (breachNotice) {
```

Comment:
This can't be null, we've already used it earlier - you can remove this if statement